### PR TITLE
Adding docstrings for SessionBase class, mode specific sesssion classes, frame and signal read and write functions

### PR DESCRIPTION
**Pull request by jashnani**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nixnet-python/blob/master/CONTRIBUTING.rst).
- [x] New tests have been created for any new features or regression tests for bugfixes.
- [ ] `tox` successfully runs, including unit tests and style checks (see [CONTRIBUTING.md](https://github.com/ni/nixnet-python/blob/master/CONTRIBUTING.rst)).

### What does this Pull Request accomplish?

Adds docstrings for the Session class. The documentation is based off of the C API documentation fouind here: http://zone.ni.com/reference/en-XX/help/372841N-01/TOC63.htm

### Why should this Pull Request be merged?

Provides meaningful documentation for the Session API. We will use this to generate html version of the docs as well.

### What testing has been done?

None

### CODE REVIEW

File: nixnet/_enums.py [hunk 1]
```diff
@@ -1093,6 +1093,71 @@ class ObjectClass(enum.Enum):
...
+            identifies a signal within the database given:
+                -<Signal>
+                -<Frame>.<Signal>
+                -<Cluster>.<Frame>.<Signal>
+                -<PDU>.<Signal>
+                -<Cluster>.<PDU>.<Signal>
+            List may also contain one or more trigger signals. For information
+            about trigger signals, refer to Signal Output Single-Point Mode or
+            Signal Input Single-Point Mode.
+
```
Comment (epage) [inline nixnet/_enums.py]: While I think its useful to move some of the documentation from the parameter descriptions into the enum docstrings, I feel this is a bit too much with the fact that it is covering how individual parameters behave rather than general impact.

File: nixnet/session.py [hunk 2]
```diff
@@ -25,6 +25,35 @@
 
 
 class Session(object):
+    """What is a Session?
+
+    The NI-XNET session represents a connection between your National Instruments
+    CAN/FlexRay/LIN hardware and hardware products on the external network.
+
```
Comment (epage) [inline nixnet/session.py]: The first-line of a docstring should be a summary that stands on its own.  "What is a Session?" does not meet that criteria

File: nixnet/session.py [hunk 3]
```diff
@@ -25,6 +25,35 @@
...
+    Each session configuration includes:
+        Interface: This specifies the National Instruments hardware to use.
+        Database objects: These describe how external hardware communicates.
+        Mode: This specifies the direction and representation of I/O data.
+
```
Comment (epage) [inline nixnet/session.py]: A session doesn't contain these but is initialized with them.  This belongs in the `__init__` docstring

File: nixnet/session.py [hunk 4]
```diff
@@ -33,7 +62,32 @@ def __init__(
             list,
             interface,
             mode):
-        "http://zone.ni.com/reference/en-XX/help/372841N-01/nixnet/nxcreatesession/"
+        """Create an XNET session at run time using strings.
```
Comment (epage) [inline nixnet/session.py]: "using strings" isn't accurate (type wise).  I assume its referring to using named references to database objects rather than receiving database references (in contrast to `nxCreateSessionByRef`).

File: nixnet/session.py [hunk 5]
```diff
@@ -33,7 +62,32 @@ def __init__(
...
+        Args:
+            database_name: The XNET database to use for interface configuration.
+                The database name must use the <alias> or <filepath> syntax
+                (refer to Databases).
+            cluster_name: The XNET cluster to use for interface configuration.
+                The name must specify a cluster from the database given in the
+                database_name parameter. If it is left blank, the cluster is
+                extracted from the list parameter; this is not allowed for modes
+                of :class:`nixnet._cconsts.NX_MODE_FRAME_IN_STREAM` or
+                :class:`nixnet._cconsts.NX_MODE_FRAME_OUT_STREAM`.
```
Comment (epage) [inline nixnet/session.py]: Why are the constants and not enum values being references?

Those constants are an implementation detail and we should act like they don't exist in public-facing documentation.

File: nixnet/session.py [hunk 6]
```diff
@@ -83,7 +137,28 @@ def __repr__(self):
         return 'Session(handle={0})'.format(self._handle)
 
     def close(self):
-        "http://zone.ni.com/reference/en-XX/help/372841N-01/nixnet/nxclear/"
+        """Close (clear) the XNET session.
+
+        This function stops communication for the session and releases all
+        resources the session uses. It internally calls
+        :class:`nixnet.nx.Session.stop` with normal scope, so if this is the
```
Comment (epage) [inline nixnet/session.py]: its no longer `nixnet.nx.Session.stop` but `nixnet.session.Session.stop`

File: nixnet/_enums.py [hunk 7]
```diff
@@ -1093,6 +1093,71 @@ class ObjectClass(enum.Enum):
```
Comment (epage) [inline nixnet/_enums.py]: Are the session properties, `types`, `errors`, and `constants` being documented separately?

File: nixnet/_session/frames.py [hunk 8]
```diff
@@ -30,14 +30,43 @@ def read_bytes(
...
+                avoid this problem, the read always returns complete frames in
+                buffer.
+
+                If 'timeout' is positive, this function waits for
+                'number_of_bytes_to_read' frame bytes to be received, then
+                returns complete frames up to that number. If the bytes do not
+                arrive prior to the 'timeout', an error is returned.
+
+                If 'timeout' is negative, this function waits indefinitely for
+                'number_of_bytes_to_read' frame bytes.
```
Comment (epage) [inline nixnet/_session/frames.py]: We should probably document this as timeout is `nixnet.constants.TIMEOUT_INFINITE`

File: nixnet/_session/frames.py [hunk 9]
```diff
@@ -30,14 +30,43 @@ def read_bytes(
...
+
+                If 'timeout' is zero, this function does not wait and immediately
+                returns all available frame bytes up to the limit
+                'number_of_bytes_to_read' specifies.
```
Comment (epage) [inline nixnet/_session/frames.py]: We should probably document this as timeout is `nixnet.constants.TIMEOUT_NONE`

File: nixnet/_session/frames.py [hunk 10]
```diff
@@ -46,14 +75,17 @@ def read_raw(
...
-        Frame: one or more
-        http://zone.ni.com/reference/en-XX/help/372841N-01/nixnet/nxreadframe/
+        """Read raw CAN frames.
+
+        Args:
+            number_to_read: An integer repesenting the number of raw CAN frames
+                to read.
+            timeout: The time to wait for number to read frames to become
+                available; the 'timeout' is represented as 64-bit floating-point
+                in units of seconds.
```
Comment (epage) [inline nixnet/_session/frames.py]: Why not talk about `NONE`/`INFINITE`?

(same for other timeouts)

File: nixnet/_session/frames.py [hunk 11]
```diff
@@ -145,14 +176,61 @@ def write_bytes(
...
+
+                If 'timeout' is negative, this functions waits indefinitely for
+                space to become available in queues.
+
+                If 'timeout' is 0, this function does not wait and immediately
+                returns with a 'timeout' error if all data cannot be queued.
+                Regardless of the 'timeout' used, if a 'timeout' error occurs,
+                none of the data is queued, so you can attempt to call this
+                function again at a later time with the same data.
+
```
Comment (epage) [inline nixnet/_session/frames.py]: ditto about -1 and 0 using constants

File: nixnet/_session/frames.py [hunk 12]
```diff
@@ -177,13 +278,30 @@ def __repr__(self):
...
 
     def write_raw(
             self,
             raw_frames):
-        "http://zone.ni.com/reference/en-XX/help/372841N-01/nixnet/nxwriteframe/"
+        """Write raw CAN frame data.
+
+        Args:
+            raw_frames: One or more :class:`nixnet.types.RawFrame` objects to be
+                written to the session.
```
Comment (epage) [inline nixnet/_session/frames.py]: Why not just say "a list of " rather than "One or more"?

The latter implies we are doing type checking for an individual item

File: nixnet/_session/frames.py [hunk 13]
```diff
@@ -193,7 +311,15 @@ def write_raw(
     def write_can(
             self,
             can_frames):
-        "http://zone.ni.com/reference/en-XX/help/372841N-01/nixnet/nxwriteframe/"
+        """Write CAN frame data.
+
+        Args:
+            can_frames: One or more :class:`nixnet.types.CanFrame` objects to be
+                written to the session.
```
Comment (epage) [inline nixnet/_session/frames.py]: ditto

File: nixnet/_session/signals.py [hunk 14]
```diff
@@ -22,14 +22,13 @@ def __repr__(self):
...
-        - Optional in C API
-        - Timestamp per data point
-        http://zone.ni.com/reference/en-XX/help/372841N-01/nixnet/nxreadsignalsinglepoint/
+        """Read data from a Signal Input Single-Point session.
+
+        Args:
+            None
+
+        Yields:
+            A tuple of timestamp (class:`nixnet._ctypedefs.nxTimestamp_t`) and signal values (float).
```
Comment (epage) [inline nixnet/_session/signals.py]: Are timestamps not integers?

Drat, we'll want to do that

File: nixnet/session.py [hunk 15]
```diff
@@ -22,22 +22,76 @@
...
     "FrameInQueuedSession",
     "FrameOutQueuedSession",
+    "FrameInSinglePointSession",
+    "FrameOutSinglePointSession",
     "SignalInSinglePointSession",
     "SignalOutSinglePointSession"]
 
 
 class SessionBase(object):
+    """Session base object.
```
Comment (epage) [inline nixnet/session.py]: This should not be public and doesn't need the same level of documentation; we should move this base class, maybe to `_session/base.py`

Alternatively, if this is truly useful for all leaf classes (including conversion), then it might be kept, just remove the reference to this being a session base class.

File: nixnet/session.py [hunk 16]
```diff
@@ -174,17 +444,34 @@ class FrameInStreamSession(SessionBase):
```
Comment (epage) [inline nixnet/session.py]: Why aren't the `Session` classes documented?

File: nixnet/_enums.py (hunk 7 above)
Comment (jashnani) [inline nixnet/_enums.py]: Yes, these will be addressed in a separate PR.

File: nixnet/_enums.py (hunk 1 above)
Comment (jashnani) [inline nixnet/_enums.py]: Since now we've created mode specific session classes, I've moved this documentation to the corresponding constructor. Ditto for other documentation related to frame/signal list syntax.

File: nixnet/session.py (hunk 2 above)
Comment (jashnani) [inline nixnet/session.py]: Fixed.

File: nixnet/session.py (hunk 3 above)
Comment (jashnani) [inline nixnet/session.py]: Fixed.

File: nixnet/session.py (hunk 4 above)
Comment (jashnani) [inline nixnet/session.py]: Fixed the comment.

File: nixnet/session.py (hunk 5 above)
Comment (jashnani) [inline nixnet/session.py]: Fixed.

File: nixnet/session.py (hunk 6 above)
Comment (jashnani) [inline nixnet/session.py]: Fixed this and other similar references.

File: nixnet/_session/frames.py (hunk 8 above)
Comment (jashnani) [inline nixnet/_session/frames.py]: Fixed.

File: nixnet/_session/frames.py (hunk 10 above)
Comment (jashnani) [inline nixnet/_session/frames.py]: Fixed.

File: nixnet/_session/frames.py (hunk 11 above)
Comment (jashnani) [inline nixnet/_session/frames.py]: Fixed.

File: nixnet/_session/frames.py (hunk 12 above)
Comment (jashnani) [inline nixnet/_session/frames.py]: Fixed.

File: nixnet/_session/frames.py (hunk 13 above)
Comment (jashnani) [inline nixnet/_session/frames.py]: Fixed.

File: nixnet/session.py (hunk 16 above)
Comment (jashnani) [inline nixnet/session.py]: Fixed.

File: nixnet/_session/frames.py (hunk 9 above)
Comment (jashnani) [inline nixnet/_session/frames.py]: Fixed.

File: nixnet/session.py (hunk 15 above)
Comment (jashnani) [inline nixnet/session.py]: I've moved the code to _session/base.py and updated the documentation as needed.
